### PR TITLE
BUG: Fix multi-index colname references in read_csv c engine.

### LINKED ISSUE #42446: option dtype in pandas.read_csv does not work properly for mulilevel columns
- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the latest version of pandas.

- [X] (optional) I have confirmed this bug exists on the master branch of pandas. 1.2.2

---

**Note**: Please read [this guide](https://matthewrocklin.com/blog/work/2018/02/28/minimal-bug-reports) detailing how to provide the necessary information for us to reproduce your bug.

#### Code Sample, a copy-pastable example
import pandas as pd
df=  pd.DataFrame({
    ('A', 'int16'):   pd.Series([1, 2, 3, 4], dtype='int16'),
    ('A', 'int32'):   pd.Series([132, 232, 332, 432], dtype='int32'), 
    ('B', 'float32'): pd.Series([1.01, 1.02, 1.03, 1.04], dtype='float32'),
    ('B', 'int16'):   pd.Series([21, 22, 23, 24], dtype='int16')})
print(df)
df.to_csv('test_df.csv')
print(df.dtypes)
# full column name tuples with level 0/1 labels don't work
df_new= pd.read_csv(
    'test_df.csv',
    header=list(range(2)),
    dtype = {
        ('A', 'int16'): 'int16', 
        ('A', 'int32'): 'int32'
    })
print(df_new.dtypes)

See my SO article for more detailed infos:
[https://stackoverflow.com/questions/54699527/dtype-is-ignored-when-using-multilevel-columns](url)

#### Problem description
Although the data types where passed in read_csv, they are not applied. dtype in read_csv only seems to work for column names that contain only one level. For Multilevel columns it generally does not seem to work.

#### Expected Output
Unnamed: 0_level_0  Unnamed: 0_level_1      int64
A                   int16                   int16
                    int32                   int32
B                   float32               float64
                    int16                   int64

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit           : 7d32926db8f7541c356066dcadabf854487738de
python           : 3.7.9.final.0
python-bits      : 64
OS               : Linux
OS-release       : 5.4.0-77-generic
Version          : #86-Ubuntu SMP Thu Jun 17 02:35:03 UTC 2021
machine          : x86_64
processor        : x86_64
byteorder        : little
LC_ALL           : None
LANG             : de_DE.UTF-8
LOCALE           : de_DE.UTF-8

pandas           : 1.2.2
numpy            : 1.20.1
pytz             : 2021.1
dateutil         : 2.8.1
pip              : 21.0.1
setuptools       : 52.0.0.post20210125
Cython           : 0.29.23
pytest           : 6.2.2
hypothesis       : None
sphinx           : 4.0.1
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : 4.5.0
html5lib         : 1.1
pymysql          : None
psycopg2         : 2.8.6 (dt dec pq3 ext lo64)
jinja2           : 2.11.3
IPython          : 7.22.0
pandas_datareader: None
bs4              : 4.9.3
bottleneck       : None
fsspec           : None
fastparquet      : None
gcsfs            : None
matplotlib       : 3.3.4
numexpr          : 2.7.3
odfpy            : None
openpyxl         : None
pandas_gbq       : None
pyarrow          : 3.0.0
pyxlsb           : None
s3fs             : None
scipy            : 1.6.2
sqlalchemy       : None
tables           : 3.6.1
tabulate         : None
xarray           : None
xlrd             : None
xlwt             : None
numba            : 0.51.2

</details>

### PR DESCRIPTION
**Pull request by jmcomie**
This fixes an issue with the read_csv c engine when the input has more than one header row and arguments to dtype, na_values, or converters reference multi-index column names as tuples.

- [x] closes #42446
- [x] tests added / passed
- [x] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing.html#code-standards) for how to run them
- [x] whatsnew entry

I added a whatsnew entry for v1.4.0 since that's what other recent bug fixes have done, but I'm not yet sure of the criteria for that. Is it appropriate to add it for whichever release "owns" master?

### CODE REVIEW

**Review by jreback — CHANGES_REQUESTED** [review]
looks fine, some comments

**Comment by jreback** [comment]
cc @gfyoung if you'd have a look

File: pandas/tests/io/parser/dtypes/test_dtypes_basic.py [hunk 1]
```diff
@@ -257,3 +257,28 @@ def test_dtype_mangle_dup_cols_single_dtype(all_parsers):
     result = parser.read_csv(StringIO(data), dtype=str)
     expected = DataFrame({"a": ["1"], "a.1": ["1"]})
     tm.assert_frame_equal(result, expected)
+
+
+def test_dtype_multi_index(all_parsers):
+    parser = all_parsers
```
[the remark this answers is not in the document]
Comment (jmcomie) [reply pandas/tests/io/parser/dtypes/test_dtypes_basic.py:264]: > can you add the issue number as a comment

done

File: pandas/_libs/parsers.pyx [hunk 2]
```diff
@@ -1281,7 +1281,10 @@ cdef class TextReader:
                 if j >= len(self.header[0]):
                     return j
                 else:
-                    return self.header[0][j]
+                    if self.has_mi_columns:
```
[the remark this answers is not in the document]
Comment (jmcomie) [reply pandas/_libs/parsers.pyx]: done

**Review by gfyoung — APPROVED** [review]

**Review by jreback — APPROVED** [review]

**Comment by jmcomie** [comment]
Regarding the failed check -- I've run the test _tests/scalar/timedelta/test_arithmetic.py_ and it's passing locally.  If it's an issue in my branch I will work to fix it but it seems this might be an environmental error ?

**Comment by jreback** [comment]
thanks @jmcomie 

failed check is not related

**Pull request merged**